### PR TITLE
kernel: rewrite SCR_SIFT_HELPER using C++ templates

### PR DESCRIPTION
**Pull request by fingolfin**
This also fixes a bug where it called DEG_PERM2 on a T_PERM4. I do not know if this caused user visible problems.

### CODE REVIEW

File: src/permutat.cc [hunk 1]
```diff
@@ -2649,150 +2649,109 @@ static Obj FuncMappingPermListList(Obj self, Obj src, Obj dst)
...
+
+    /* Copy g into the buffer */
+    Res * ptR = ADDR_PERM<Res>(result);
+    const TG * ptG = CONST_ADDR_PERM<TG>(g);
+    if (sizeof(TG) == sizeof(Res)) {
+        memcpy(ptR, ptG, sizeof(TG) * dg);
     }
     else {
-        const UInt4* ptrResult = CONST_ADDR_PERM4(result);
-        UInt degResult = DEG_PERM2(result);
```
Comment (fingolfin) [inline src/permutat.cc]: @stevelinton this is the incorrect `DEG_PERM2` call.

Also, I wonder: isn't the value returned by DEG_PERM4 here equal to `nn` anyway?

File: src/permutat.cc [hunk 2]
```diff
@@ -2649,150 +2649,109 @@ static Obj FuncMappingPermListList(Obj self, Obj src, Obj dst)
...
+}
+
+static Obj FuncSCR_SIFT_HELPER(Obj self, Obj stb, Obj g, Obj n)
+{
+    if (!IS_PREC(stb))
+        RequireArgument("SCR_SIFT_HELPER", stb, "must be a plain record");
+    RequirePermutation("SCR_SIFT_HELPER", g);
+    UInt nn = GetSmallInt("SCR_SIFT_HELPER", n);
+
+    if (!RN_stabilizer)
```
Comment (ChrisJefferson) [inline src/permutat.cc]: Seems a bit weird to put these in here, instead of SCR_SIFT_HELPER, which actually needs/uses them?

Comment (fingolfin) [reply src/permutat.cc]: This is shared code, so not putting it in there reduces the size of the generated code. Also, IMO it feels a bit less cluttered... In fact, strictly speaking, I'd prefer to init `RN_stabilizer` in `PostRestore`... I'll adjust the PR.

**Comment by ChrisJefferson** [comment]
I already approved, but just for confirmation, I reapprove :)

**Comment by fingolfin** [comment]
The test failure exposed by this PR is fixed by #3245

**Pull request merged**